mlflow/genai/utils/trace_utils.py

```diff
@@ -837,3 +837,385 @@ def _parse_tools_to_chat_tool(tools_data: list[Any]) -> list[ChatTool]:
+            content = message.get("content", "")
+            if tool_call_id:
+                tool_message_content_map[tool_call_id] = content
+    return tool_message_content_map
+
+
+def _normalize_conversation(data: Any, message_format: str | None) -> list[dict[str, Any]]:
+    """
+    Normalize conversation input data based on message format.
+
+    Mirrors frontend's normalizeConversation function (ModelTraceExplorer.utils.tsx).
+    Routes to format-specific parsers to extract messages with tool_calls from input data.
+
+    Args:
+        data: Span inputs data
+        message_format: Format identifier from SpanAttributeKey.MESSAGE_FORMAT
+
+    Returns:
+        List of normalized message dictionaries with tool_calls
+    """
+    if not data:
+        return []
+
+    try:
+        # Route to format-specific parsers
+        if message_format == "openai":
+            return _normalize_openai_chat_input(data)
+        elif message_format == "openai-agent":
+            return _normalize_openai_agent_input(data)
+        elif message_format == "langchain":
+            return _normalize_langchain_chat_input(data)
+        else:
+            # Fallback: try OpenAI format
+            return _normalize_openai_chat_input(data)
+    except Exception:
+        return []
+
+
+def _normalize_openai_chat_input(data: Any) -> list[dict[str, Any]]:
+    """
+    Parse OpenAI input format to extract messages with tool_calls.
+
+    Mirrors: normalizeOpenAIChatInput (chat-utils/openai.ts)
+
+    Handles OpenAI input formats:
+    - {messages: [...]}
+    - {input: [...]}
+    - Direct array of messages
+    """
+    # Handle INPUT format: {messages: [...]} or {input: [...]}
+    if isinstance(data, dict):
+        input_messages = data.get("messages") or data.get("input")
+        if input_messages and isinstance(input_messages, list):
+            return [msg for msg in input_messages if isinstance(msg, dict) and "role" in msg]
+
+    # Handle direct array of messages
+    if isinstance(data, list) and all(isinstance(msg, dict) and "role" in msg for msg in data):
+        return data
+
+    return []
+
+
+def _normalize_openai_agent_input(data: Any) -> list[dict[str, Any]]:
+    """
+    Parse OpenAI Agent input format to extract messages with tool_calls.
+
+    Mirrors: normalizeOpenAIAgentInput (chat-utils/openai.ts)
+
+    Handles agent format where function calls and messages are in an array:
+    - {type: "function_call", call_id, name, arguments}
+    - {type: "function_call_output", call_id, output}
+    - Regular messages with {role, content}
+    """
+    if not isinstance(data, list) or len(data) == 0:
+        return []
+
+    # Check if all items are valid OpenAI Agent messages
+    if not all(_is_openai_agent_message(item) for item in data):
+        return []
+
+    return [msg for item in data if (msg := _normalize_openai_agent_message(item))]
+
+
+def _is_openai_agent_message(obj: Any) -> bool:
```

Comment:
can we create a base interface for all these types of messages and have each provider implement the _is_applicable, _normalize_message, etc? similar to how we do it in mlflow.genai.judges.adapters